datasource/ethereum/src/block_ingestor.rs

```diff
@@ -0,0 +1,136 @@
+use failure::Error;
+use std::fmt::Debug;
+use std::sync::Mutex;
+use std::time::Duration;
+use std::time::Instant;
+use web3::api::Web3;
+use web3::transports::batch::Batch;
+use web3::types::Block;
+use web3::types::BlockId;
+use web3::types::BlockNumber;
+use web3::types::H256;
+use web3::types::Transaction;
+use web3::BatchTransport;
+
+use graph::prelude::*;
+
+#[derive(Clone, Debug)]
+pub struct BlockIngestorConfig<S: BlockStore, T: BatchTransport> {
+    pub store: Arc<Mutex<S>>,
+    pub network_name: String,
+    pub web3_transport: T,
+    pub ancestor_count: u64,
+    pub logger: slog::Logger,
+}
+
+pub struct BlockIngestor<S: BlockStore, T: BatchTransport + Debug + Clone> {
+    store: Arc<Mutex<S>>,
+    network_name: String,
+    web3_transport: T,
+    ancestor_count: u64,
+    logger: slog::Logger,
+}
+
+impl<S: BlockStore + Send + 'static, T: BatchTransport + Debug + Clone> BlockIngestor<S, T> {
+    pub fn spawn(config: BlockIngestorConfig<S, T>) -> Result<(), Error>
+    where
+        T: Send + Sync + 'static,
+    {
+        // Extract config params
+        let BlockIngestorConfig {
+            store,
+            network_name,
+            web3_transport,
+            ancestor_count,
+            logger,
+        } = config;
+
+        // Add a head block pointer for this network name if one does not already exist
+        store.lock().unwrap().add_network_if_missing(&network_name)?;
+
+        // Start block ingestor
+        let block_ingestor = BlockIngestor {
+            store,
+            network_name,
+            web3_transport,
+            ancestor_count,
+            logger: logger.new(o!("component" => "BlockIngestor")),
+        };
+        tokio::spawn(block_ingestor.into_polling_stream());
+
+        Ok(())
+    }
+
+    fn into_polling_stream(self) -> impl Future<Item = (), Error = ()> {
+        let err_logger = self.logger.clone();
+
+        tokio::timer::Interval::new(Instant::now(), Duration::from_secs(5))
```

Comment:
Where does this 5s come from?

Reply:
Ha, that's the polling interval. Shall I make that a command line arg?

Reply:
5s is kind of an eternity. Maybe 1s or less.. Ideally we'd be able to process a block and push any changes in <500ms.

Reply:
Oh I see!
I just pushed a new commit to add `polling_interval` to the `BlockIngestorConfig` and reduce it to 0.5s. Also created #310 for config file.